Fix #8423: Added a lint check to prevent usage of string concatenation, and promote string interpolation instead.

tanishq67 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #8423 .
2. This PR does the following: This PR contains a lint check to prevent usage of string concatenation, and promote string interpolation instead.

**Accepted standards:**
```
a = '%s%s' % ('hello','mello')

b = '%smello' % a

c = 'hello%s' % b
```
**Pylint Error:**
```
a = 'hello' + 'mello'

b = a + 'mello' 

c = 'hello' + b

```
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

tanishq67 committed:
Add a lint check to prevents usage of string concatenation, and promotes string interpolation instead.

tanishq67 committed:
Add a lint check to prevents usage of string concatenation, and promotes string interpolation instead.

tanishq67 committed:
Add a lint check to prevents usage of string concatenation, and promotes string interpolation instead.

tanishq67 commented:
@sajalasati  PTAL

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1929,6 +1931,47 @@ def visit_functiondef(self, node):
 
+class ConcatenationChecker(checkers.BaseChecker):
+    """Custom pylint checker which checks that string
+    concatenation is present or not
```
Please end the docstring with a fullstop.
@DubeySandeep Do you think this should be caught by some linter check?

DubeySandeep replied:
I think we have such a linter check, If we don't have one then we should add one!

sajalasati replied:
@DubeySandeep Thanks! I have added it as a task in #8423 now.

tanishq67 replied:
Done

sajalasati replied:
I don't see the change yet. Also rephrasing it to something like `Custom pylint checker which prohibits use of string concatenation and promotes string interpolation instead.` will make it sound better and clearer.

tanishq67 replied:
Done!!!

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -1929,6 +1931,47 @@ def visit_functiondef(self, node):
+    priority = -1
+    msgs = {
+        'C0032': (
+            'At string %s, please prefer string interpolation over '
+            'concatenation',
+            'string-concatenation',
+            'Prefer: "My string %s" % varname to "My string " + varname.',
```
Rephrase the 1st and 3rd message such that the 1st message explains what a user needs to do to fix the issue in the codebase, and the 3rd message should be info about this lint check i.e, what this lint check provides? See [doc](http://pylint.pycqa.org/en/latest/how_tos/custom_checkers.html) for more info on what each of it means. Treat other checkers in this file as an example.
For eg 3rd msg could be like: `Enforce use of string interpolation over string concatenation.`

tanishq67 replied:
Done

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -2977,3 +2977,48 @@ def some_random_function(param1, param2):
+            pylint_extensions.ConcatenationChecker)
+        self.checker_test_object.setup_method()
+
+    def test_string_concatenation(self):
```
There should be a test method for each kind of case - like here atleast two are needed - one for the case where error is raised, and other where no error should be raised. Also we follow a standard convention for naming the test methods - you can find info about both these things in our [guide](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) to help you structure the tests in a better way.

tanishq67 replied:
Done

sajalasati reviewed (changes requested):
@tanishq67 Thanks for your PR! I did a high-level review and left some comments for you. Please resolve the merge conflicts first as per oppiabot's msg, and I will review it detail once the conflict is resolved and the comments have been addressed.

DubeySandeep commented:
@tanishq67, Can you please resolve the merge conflicts? [Feel free to assign reviewers once the PR is ready for the review :)]

tanishq67 committed:
Resolved merge conflicts

tanishq67 commented:
@DubeySandeep @sajalasati  PTAL

tanishq67 commented:
I resolved all the merge conflicts and implemented the comments you wrote. I also renamed variables following the convention- ``` test_{{action}}_{{withCondition1}}_{{withCondition2}}_{{hasExpectedOutcome}}```. While reading the testing documentation, I came to realize that there is  one edge case which might fail in my Lint check, so I revamped the logic completely, and wrote a better one.

DubeySandeep commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,56 @@ def visit_call(self, node):
+        ),
+    }
+
+    def process_tokens(self, tokens):
```
Instead of the parsing tokens, can we do it through astroid/nodes? [For [ref](https://docs.google.com/document/d/1v0Oi2qFIC3HXwFaCX1zpG8aP0IjtZnYs_s_U8S-FHn8/edit?usp=sharing)]

I mean will it be possible to visit operators and if the operator is '+' and either of the left or right values are string then raise lint issue? (not sure!)

tanishq67 replied:
**Done**

tanishq67 replied:
Used astroid/nodes this time.

DubeySandeep reviewed (changes requested):
@tanishq67 Thanks for the PR, I've left one general comment, PTAL!

DubeySandeep commented:
@tanishq67, The lint checks are failing on this PR. Note: You may have to fix the existing issue in the codebase which caught-up by the new lint check introduced in this PR!

tanishq67 commented:
Hello, thanks for the comments @DubeySandeep.
While fixing the existing issue in the codebase which were caught-up by the new lint check, I came to realize that they are about 300+ errors. Should I fix them manually ? or if possible can you give me ideas to automate it.

sajalasati commented:
> Hello, thanks for the comments @DubeySandeep.
> While fixing the existing issue in the codebase which were caught-up by the new lint check, I came to realize that they are about 300+ errors. Should I fix them manually ? or if possible can you give me ideas to automate it.

@tanishq67 I think you could Python's regex methods for automating that, something like `re.sub()` [ref](https://docs.python.org/2/library/re.html#re.sub)  (I am not sure if that alone would help you. You can google about it more). You might find fix it manually to be easier in my opinion though.

sajalasati commented:
Hi @tanishq67! Any updates on the PR? If you are stuck at some specific problem(s), do let us know here. Thanks!

tanishq67 commented:
@sajalasati really sorry for the delay, I was busy with my mid-semester examinations, just yesterday evening my midsems ended.

tanishq67 committed:
Changed implementation from token_checker to astroid_checker and fixed all lint errors in the existing codebase.

tanishq67 commented:
PTAL @sajalasati @DubeySandeep

DubeySandeep commented:
@sajalasati, Can you please take the first pass? :)

vojtechjelinek reviewed (approved):
LGTM!

BenHenning reviewed (approved):
Approved for my codeowners files.

nithusha21 reviewed (approved):
LGTM for codeowner files!

aks681 reviewed (approved):
lgtm as codeowner

sajalasati commented:
@tanishq67 Can you first reply to all the previous comments with a msg like "DONE" or something? (Make sure you always do this whenever requesting/re-requesting for review, because it helps us know you have addressed our concerns. Keeping track of your changes just from comments like [these](https://github.com/oppia/oppia/pull/11910#issuecomment-780298249) becomes difficult over time). All of this is also mentioned in our [contributor guide](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me).

I'll keep myself assigned and review as soon as you do this. Also resolve the merge conflict. Thanks!

tanishq67 commented:
@sajalasati  Thank you very much for the guidance. I have addressed all the previous comments.

tanishq67 commented:
@sajalasati I will handle all the merge conflicts after I implement changes you will suggest.

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,42 @@ def visit_call(self, node):
+    }
+
+    def visit_binop(self, node):
+        """Custom pylint checker which makes sure that string concatenation
```
Since this the docstring of the function, we can be more specific about what the function does. See [this](https://github.com/oppia/oppia/blob/3814a20931432f18faf88fc027840292d2b1ed57/scripts/linters/pylint_extensions.py#L2078) or [this](https://github.com/oppia/oppia/blob/3814a20931432f18faf88fc027840292d2b1ed57/scripts/linters/pylint_extensions.py#L1926) for better idea. (What you have written currently can go as the docstring of the checker class)

tanishq67 replied:
Thank you very much for the suggestions, I have just changed it please take a look

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,42 @@ def visit_call(self, node):
+        """
+        if node.op == b'+':
+            for operand in node.get_children():
+                if (isinstance(operand, astroid.nodes.Const) and
+                        ('str' in operand.pytype())):
```
Just listing down all the cases we want to cover
1. Both are strings: `'str1'+'str2'`
2. One of them is a string, and the other one is a variable that stores a string: `'str1'+var` and `var+'str1'`
3. Both are variables that are store a string (@DubeySandeep we should be covering this case too, right?)

If we are covering 3rd possibility also, then I don't think this code covers that, right @tanishq67? You should make changes accordingly.

sajalasati replied:
Sorry my bad. String interpolation doesn't make sense when both operands are variables, so case 3 is not possible.

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,48 @@ def some_random_function(param1, param2):
+            pylint_extensions.ConcatenationChecker)
+        self.checker_test_object.setup_method()
+
+    def test_string_concatenation_present_error(self):
```
For naming test methods please use a descriptive name that also follows the [convention](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests) I told you earlier to look in the guide - Note that it has 3 parts: action, with condition, outcome, so you should include all whenever possible. Like I would have named this test as `test_string_concatenation_with_string_operands_raises_error` - always make it descriptive and if it exceeds 80 char limit, maybe drop few words or make them shorter. 

You can also split this test for multiple cases (which you have considered) if you want, because each test should be testing only one kind of functionality or behaviour (read about this too in the same guide).

tanishq67 replied:
I did read the wiki and tried to follow it but the phrase I was using was exceeding the 80 word limit, I really liked your suggestion and will implement them. From next time onwards I will definitely carefully ponder over nomenclature. Thank you

tanishq67 replied:
Wrote better names this time! added ur suggestions as well as a complex test.

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,48 @@ def some_random_function(param1, param2):
+            self.checker_test_object.checker.visit_binop(binop_node2)
+            self.checker_test_object.checker.visit_binop(binop_node3)
+
+    def test_string_concatenation_not_present(self):
```
Better naming of the test: `test_string_interpolation_does_not_raise_error`. 

Perhaps we should also have:
- A test method to test the behaviour in presence of multiple `+` symbols. Just take any one of the 3 cases, and verify the checker works as expected.
- Single test method to test the behaviour in other scenarios where its neither the case of string interpolation, nor string concatenation but has '+' operator. For eg. we are adding 2 non-string variables, `a+b` - this should not raise any error.

If you feel there are other such test cases too (both where is/isn't raised), please add tests for them too.

tanishq67 replied:
Oh, I was naming it something similar but 80 word limit was being exceeded so, I cutoff words to fit it. Your suggestions are really helpful, I will implement them and more quality tests as well. Thank you very much for the guidance.

tanishq67 replied:
Done!

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,42 @@ def visit_call(self, node):
                     break
 
 
+class ConcatenationChecker(checkers.BaseChecker):
```
I think `StringConcatenationChecker` would be more apt, what do you think?

tanishq67 replied:
Oh thank you for pointing it out, yes definitely it will be better.

tanishq67 replied:
Done!

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,42 @@ def visit_call(self, node):
+            for operand in node.get_children():
+                if (isinstance(operand, astroid.nodes.Const) and
+                        ('str' in operand.pytype())):
+                    self.add_message(
```
Why are we adding a message for every operand? Why not just generate the message once per `+` operator?

tanishq67 replied:
Thank you for the suggestion, but I think this case is being handled by the break statement in the next line.

sajalasati replied:
Ah that's true but partially. Let me re-explain my question properly with the additional observation I have now.

Looking at one of your tests `test_multiple_addition_sign_with_string_operands_raises_error `, I tested out that case in real also, and felt the error msgs produce a slight confusion which is linked with the approach. To understand if we can do better and **prevent multiple error msgs for a single string concatenation expression**, I did an inspection of how astroid parsing happens for binop. Say we have a general expression with multiple `+` operators like below:
```python
'string_1'+'string_2'+'string_3'+string_var
```
then `astroid.extract_node()` starts parsing it from rightmost `+` symbol - it calls `string_var` as the right operand, and rest of the left side expression (`'string_1'+'string_2'+'string_3'`) as left operand, on which it again recurses - it will reach leftmost `+` after 2 more recursions to give `'string_1'` and `'string_2'` as its left and right children.

Suggestions for the approach:
1. Any `astroid.node.BinOp` only has (exactly) 2 children i.e left and right. So you can avoid iterating over the children and just address them as `node.left` or `node.right` for simplicity.
2. You raise error msg only for leftmost `+` operator in the expression using the fact that for all other `+`, left operand will of the type `astroid.node.BinOp` except for the leftmost `+`. Though this is also based on the assumption that at least one of those operands is string (**which you currently do assume**), which seems fair enough to me.

But just to double check @DubeySandeep @seanlip do you think we should include `a+b` in cases of prohibiting string concatenation where both `a` and `b` are strings? To me it does not make sense doing that, because then it's better to keep just one string variable. Also it's not so straightforward to cover that case, and would need more thinking.

sajalasati replied:
Please also do modifications in error msg/tests as needed.

sajalasati reviewed (changes requested):
@tanishq67 Reviewed your approach in `pylint_extensions` file, and it looks better now! Left comments to restructure and add more tests, and about some other concerns. Thanks for going over all the files and making the corrections -- leaving it to codeowners to verify them, though they look good to me on a high level.
Thanks!

sajalasati commented:
@tanishq67 After reviewing, I edited some comments so make sure you go through the updated comments only. If you have any doubts about any of those, do let me know. Thanks!

iamprayush reviewed (approved):
LGTM for code-owner file!

tanishq67 committed:
Addressed comments.

tanishq67 commented:
@sajalasati PTAL

tanishq67 commented:
@sajalasati Also please guide me to resolve the failing test

sajalasati commented on scripts/linters/pylint_extensions.py:
```diff
@@ -2109,6 +2111,43 @@ def visit_call(self, node):
+    }
+
+    def visit_binop(self, node):
+        """called to find operator '+' so that we can find string
```
The first letter should be capital, and we could improve the language also here.
My Suggestion: `Called for every '+' operator to prohibit usage of string concatenation.`

@DubeySandeep Currently we don't have lint check to ensure docstring starts with capital letter only - it makes sense to add one, right?

tanishq67 replied:
Nice suggestion, implemented it!!!

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,93 @@ def some_random_function(param1, param2):
+
+    def test_string_concatenation_with_string_operands_raises_error(self):
+        binop_node1, binop_node2, binop_node3 = astroid.extract_node(
+            """
+        'string1' + 'string2' #@
+        a + 'string1' #@
+        'string2' + b #@
+        """)
```
Move right by 4 spaces to correct the indentation as it looks a bit weird currently, i.e keep all the lines indented with 4 space shift from line where `(` starts. (If the suggested change raises lint error please let me know)
```suggestion
            """
            'string1' + 'string2' #@
            a + 'string1' #@
            'string2' + b #@
            """)
```

tanishq67 replied:
Yes, you are right !! Done

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,93 @@ def some_random_function(param1, param2):
             self.checker_test_object.checker.visit_functiondef(def_node)
+
+
+class ConcatenationCheckerTests(unittest.TestCase):
```
Change this as per Checker class name -> `StringConcatenationCheckerTests`

tanishq67 replied:
Ah, yes my bad! Thanks

tanishq67 replied:
Done!!!

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,93 @@ def some_random_function(param1, param2):
+        with self.checker_test_object.assertNoMessages():
+            self.checker_test_object.checker.visit_binop(binop_node1)
+
+    def test_multiple_addition_sign_with_string_operands_raises_error(self):
```
A better name suggestion - `test_multiple_string_concatenations_raises_error`

tanishq67 replied:
Definately it will be better.

tanishq67 replied:
Done!!!

sajalasati commented on scripts/linters/pylint_extensions_test.py:
```diff
@@ -3257,3 +3257,93 @@ def some_random_function(param1, param2):
+            self.checker_test_object.checker.visit_binop(binop_node1)
+            self.checker_test_object.checker.visit_binop(binop_node2)
+
+    def test_addition_of_non_string_variables_does_not_raise_error(self):
```
Have you tested the case when we concatenate 2 string variables?

tanishq67 replied:
Done !!!!!!!!

sajalasati reviewed (changes requested):
Thanks @tanishq67 ! The restructured tests look good now and cover mostly all cases. Looking at one of them I just realised we can change our approach to print error msgs only one per expression instead of one per every string operand/every `+`. PTAL at review for details. Left other minor nit suggestions.

Also can you update the Next message id on top of `pylint_extensions.py` file to be `C0035`.

sajalasati commented:
@brianrodri PTAL at the files owned by you! Thanks!

DubeySandeep commented:
@brianrodri Can you please review this PR?

tanishq67 commented:
@sajalasati Thank you very very much for reviewing.
Variables in python are dynamically typed. I looked up about this and found that static type checking is introduced only after `python 3.6`. Without that, I think, we cannot determine the type of a variable by static code analysis. So, it is not easy ( or maybe impossible ) to detect cases like `a+b` where **a** and **b** are string variables. 

The idea you proposed for the multiple operand case is good but I think it will fail in certain conditions.
For eg.  If we will assume that there is **atleast one string operand** in the expression then it will also give error in this case ` a+b+3+5 ` (where **a** and **b** are **integers**) which we dont want. And if we will not assume, then it will miss cases like ` a+b+'string1'+'string2' ` (where **a** and **b** are **string variables**).
I tried to think of any other better solution but I was unable to. However, still I think current solution fullfills our [basic requirements](https://github.com/oppia/oppia/issues/8423#issuecomment-774630865).
Do let me know if you have any other ideas/suggestions.

tanishq67 commented:
@sajalasati PTAL

sajalasati commented:
@tanishq67 Why did you force push? Read through our [wiki](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#address-review-comments-until-all-reviewers-give-lgtm-looks-good-to-me) again and understand it's negative consequences. Please avoid doing it.

sajalasati commented:
@tanishq67 The commit history has got entirely messed up, it says the last commit was on 10 days ago. Please create a new PR, I am closing this one (and remember never to force push in future).
